common/src/Renderer/EntityDecalRenderer.cpp

```diff
@@ -0,0 +1,438 @@
+    std::memcpy(vertDest, vertices.data(), vertices.size() * sizeof(*vertDest));
+    data.vertexHolderKey = vertBlock;
+
+    const auto brushVerticesStartIndex = static_cast<GLuint>(vertBlock->pos);
+
+    auto& faceVboMap = *m_faces;
+    auto& holderPtr = faceVboMap[texture];
+    if (holderPtr == nullptr)
+    {
+      // inserts into map!
+      holderPtr = std::make_shared<BrushIndexArray>();
+    }
+    auto [indexBlock, indexDest] =
+      holderPtr->getPointerToInsertElementsAt(indices.size());
+    auto* currentDest = indexDest;
+    for (const auto& i : indices)
+    {
+      *(currentDest++) = static_cast<GLuint>(brushVerticesStartIndex + i);
+    }
+    data.faceIndicesKey = indexBlock;
+  }
+
+  data.validated = true;
+}
+
+using Vertex = Renderer::GLVertexTypes::P3NT2::Vertex;
+std::vector<Vertex> EntityDecalRenderer::createDecalBrushFace(
+  const Model::EntityNode* entityNode,
+  const Model::BrushNode* brush,
+  const Model::BrushFace& face,
+  const Assets::Texture* texture)
+{
+  assert(texture != nullptr);
+
+  const auto textureWidth = static_cast<float>(texture->width());
+  const auto textureHeight = static_cast<float>(texture->height());
+  const auto textureSize = vm::vec2f{textureWidth, textureHeight};
+  const auto textureName = texture->name();
+
+  // copy the face properties, used to calculate the decal size and texture coords
+  auto attrs = Model::BrushFaceAttributes{textureName, face.attributes()};
+  auto texCoords = face.texCoordSystem().clone();
+  const auto tex = texCoords.get();
+
+  // create the geometry for the decal
+  const auto plane = face.boundary();
+  const auto origin = entityNode->physicalBounds().center();
+  const auto center = plane.project_point(origin);
+
+  // re-project the vertices in case the texture axes are not on the face plane
+  const auto xShift = tex->xAxis() * static_cast<double>(attrs.xScale())
+                      * static_cast<double>(textureWidth) / 2.0;
+  const auto yShift = tex->yAxis() * static_cast<double>(attrs.yScale())
+                      * static_cast<double>(textureHeight) / 2.0;
+
+  // we want to shift every vertex by just a little bit to avoid z-fighting
+  const auto offset = plane.normal * 0.1;
+
+  // start with a rectangle
+  auto verts = std::vector{
+    plane.project_point(center + xShift - yShift) + offset, // Bottom Right
+    plane.project_point(center + xShift + yShift) + offset, // Top Right
+    plane.project_point(center - xShift + yShift) + offset, // Top Left
+    plane.project_point(center - xShift - yShift) + offset  // Bottom Left
+  };
+
+  // because the texture axes don't have to align to the face, we might have a reversed
+  // face here. if so, reverse the points to get a valid face for the plane
+  // todo: is there a better way to do this?
+  const auto [_, vertPlane] = vm::from_points(verts[0], verts[1], verts[2]);
+  if (!vm::is_equal(plane.normal, vertPlane.normal, 0.01))
```

Comment:
Why `0.01` and not `vm::C::almost_equal()`?

Reply:
I didn't know that was an option!